tests/unittest_nodes_lineno.py

```diff
@@ -0,0 +1,1185 @@
+
+        c2 = ast_nodes[1]
+        assert isinstance(c2, nodes.SetComp)
+        assert isinstance(c2.elt, nodes.Name)
+        assert isinstance(c2.generators[0], nodes.Comprehension)  # type: ignore
+        assert (c2.lineno, c2.col_offset) == (2, 0)
+        assert (c2.end_lineno, c2.end_col_offset) == (2, 16)
+        assert (c2.elt.lineno, c2.elt.col_offset) == (2, 1)
+        assert (c2.elt.end_lineno, c2.elt.end_col_offset) == (2, 2)
+
+        c3 = ast_nodes[2]
+        assert isinstance(c3, nodes.DictComp)
+        assert isinstance(c3.key, nodes.Name)
+        assert isinstance(c3.value, nodes.Name)
+        assert isinstance(c3.generators[0], nodes.Comprehension)  # type: ignore
+        assert (c3.lineno, c3.col_offset) == (3, 0)
+        assert (c3.end_lineno, c3.end_col_offset) == (3, 22)
+        assert (c3.key.lineno, c3.key.col_offset) == (3, 1)
+        assert (c3.key.end_lineno, c3.key.end_col_offset) == (3, 2)
+        assert (c3.value.lineno, c3.value.col_offset) == (3, 4)
+        assert (c3.value.end_lineno, c3.value.end_col_offset) == (3, 5)
+
+        c4 = ast_nodes[3]
+        assert isinstance(c4, nodes.GeneratorExp)
+        assert isinstance(c4.elt, nodes.Name)
+        assert isinstance(c4.generators[0], nodes.Comprehension)  # type: ignore
+        assert (c4.lineno, c4.col_offset) == (4, 0)
+        assert (c4.end_lineno, c4.end_col_offset) == (4, 16)
+        assert (c4.elt.lineno, c4.elt.col_offset) == (4, 1)
+        assert (c4.elt.end_lineno, c4.elt.end_col_offset) == (4, 2)
+
+    @staticmethod
+    def test_end_lineno_class() -> None:
+        """ClassDef, Keyword."""
+        code = textwrap.dedent(
+            """
+        @decorator1
+        @decorator2
+        class X(Parent, var=42):
+            pass
+        """
+        ).strip()
+        c1 = builder.extract_node(code)
+        assert isinstance(c1, nodes.ClassDef)
+        assert isinstance(c1.decorators, nodes.Decorators)
+        assert isinstance(c1.bases[0], nodes.Name)
+        assert isinstance(c1.keywords[0], nodes.Keyword)
+        assert isinstance(c1.body[0], nodes.Pass)
+
+        assert (c1.lineno, c1.col_offset) == (3, 0)
+        assert (c1.end_lineno, c1.end_col_offset) == (4, 8)
+        assert (c1.decorators.lineno, c1.decorators.col_offset) == (1, 0)
+        assert (c1.decorators.end_lineno, c1.decorators.end_col_offset) == (2, 11)
+        assert (c1.bases[0].lineno, c1.bases[0].col_offset) == (3, 8)
+        assert (c1.bases[0].end_lineno, c1.bases[0].end_col_offset) == (3, 14)
+        if PY39_PLUS:
+            # 'lineno' and 'col_offset' information only added in Python 3.9
+            assert (c1.keywords[0].lineno, c1.keywords[0].col_offset) == (3, 16)
+            assert (c1.keywords[0].end_lineno, c1.keywords[0].end_col_offset) == (3, 22)
+        assert (c1.body[0].lineno, c1.body[0].col_offset) == (4, 4)
+        assert (c1.body[0].end_lineno, c1.body[0].end_col_offset) == (4, 8)
```

Comment:
I just reviewed the content of the test, not the expected column. But it comes from the ast, if it's wrong there isn't a lot we can do, right ?

Reply:
True, at least not without significant effort. One area I was able to fix fairly easily was `Arguments`. See the snipped below. We might also be able to address the `AssignName` issue I mentioned by parsing the source code ourselves but that is definitely an issue for another time.

Usually the worse that can happen is that the lineno and col_offset also include the parent node. This is still correct just not as accurate as we would like it to be for error highlighting.

The true value of these tests is to know when a new version changed something. Like knowing that `Keywords` and `Slice` only include `end_lineno` in 3.9, or the issue with `Arguments`.

--
https://github.com/PyCQA/astroid/blob/7dd55fe8d96846617b2827e5272582ffc865190f/astroid/rebuilder.py#L834-L840